### PR TITLE
Example cross market maker + byproduct utility functions

### PR DESCRIPTION
**Pull request by waterquarks**
- Adds an ensureOpenOrdersAccount function function, that creates an open orders account for the submitted Serum market if it doesn't already exist. This is tested against an example market maker which is to receive further updates.
- Install KrisM's risk check.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -87,6 +87,7 @@
     "buffer-layout": "^1.2.1",
     "cross-fetch": "^3.1.5",
     "dotenv": "^10.0.0",
+    "mango_risk_check": "^1.2.1",
```
Comment (mschneider) [inline package.json:90]: pls move to dev dependencies

File: src/client.ts [hunk 2]
```diff
@@ -5491,4 +5491,60 @@ export class MangoClient {
 
     return await this.sendTransaction(transaction, owner, []);
   }
+
+  async ensureOpenOrdersAccount(
+      mangoAccount: MangoAccount,
```
Comment (mschneider) [inline src/client.ts:5496]: code is not formatted. please run prettier

File: src/scripts/xmm.ts [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+
+                if (!(fill.maker.equals(mangoAccountPk))) {
+                    return
+                }
+
+                recentFills.push({...fill, slot: data.slot})
+
+                while (recentFills.length > 256) {
+                    recentFills.shift()
+                }
```
Comment (mschneider) [inline src/scripts/xmm.ts:261]: not a big fan of this, a better solution would be to just filter the array and override it:

```
recentFills = recentFills.filter(fill => fill.slot > slot)
```

File: src/scripts/xmm.ts [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+
+                const takerBase = perpMarket.baseLotsToNumber(mangoAccount.perpAccounts[perpMarketConfig!.marketIndex].takerBase)
+
+                const basePosition = perpMarket.baseLotsToNumber(mangoAccount.perpAccounts[perpMarketConfig!.marketIndex].basePosition)
+
+                const queuedBasePosition = perpEventQueue.getUnconsumedEvents()
+                    .filter(event => event.fill !== undefined)
+                    .map(event => event.fill)
+                                // @ts-ignore
```
Comment (mschneider) [inline src/scripts/xmm.ts:270]: code shouldn't ignore any typescript compilation issues

File: src/scripts/xmm.ts [hunk 5]
```diff
@@ -0,0 +1,557 @@
...
+                askId,
+                'sell',
+                new BN(255),
+                'postOnlySlide',
+                false,
+                undefined,
+                expiryTimestamp
+            ),
+            // @ts-ignore
+            // tx.add(riskChecker.makeCheckRiskInstruction(perpMarketConfig, perpMarket))
```
Comment (mschneider) [inline src/scripts/xmm.ts:407]: risk check should be enabled

File: src/scripts/xmm.ts [hunk 6]
```diff
@@ -0,0 +1,557 @@
...
+    async function monitor(): Promise<[number, MangoAccount, PerpEventQueue, BlockhashWithExpiryBlockHeight, number]> {
+        const [mangoAccountRaw, perpEventQueueRaw] = await getMultipleAccounts(
+            connection,
+            [mangoAccountPk, perpMarketConfig!.eventsKey]
+        )
+
+        const slots = _.uniq([mangoAccountRaw.context.slot, perpEventQueueRaw.context.slot])
+
+        if (slots.length !== 1) {
+            throw new Error('Inconsistent slots on Mango account monitor.')
```
Comment (mschneider) [inline src/scripts/xmm.ts:109]: this error is not correctly handled in the calling function, might cause "unhandled promise rejection" issues at runtime which can be really hard to trace, better to return an optional type

File: src/scripts/xmm.ts [hunk 7]
```diff
@@ -0,0 +1,557 @@
...
+        }
+
+        const slot = slots[0]
+
+        const mangoAccount = new MangoAccount(mangoAccountPk, MangoAccountLayout.decode(mangoAccountRaw.accountInfo.data))
+
+        const perpEventQueue = new PerpEventQueue(PerpEventQueueLayout.decode(perpEventQueueRaw.accountInfo.data))
+
+        const recentBlockHash = await connection.getLatestBlockhash('finalized')
```
Comment (mschneider) [inline src/scripts/xmm.ts:118]: should be in parallel with getMultipleAccounts

File: src/scripts/xmm.ts [hunk 8]
```diff
@@ -0,0 +1,557 @@
...
+        // ^ Solana transactions require a recent block hash passed as metadata in order to be signed.
+        // Instead of fetching this before dispatching every transaction (which would delay it by a
+        // few milliseconds) we poll for this regularly in the background.
+
+        const recentBlockTime = await connection.getBlockTime(
+            await connection.getSlot('finalized')
+        )
```
Comment (mschneider) [inline src/scripts/xmm.ts:125]: should be in parallel with getMultipleAccounts

File: src/scripts/xmm.ts [hunk 9]
```diff
@@ -0,0 +1,557 @@
...
+            // eslint-disable-next-line no-prototype-builtins
+            const isSnapshot = data.hasOwnProperty('events')
+
+            const parseEvent = (event: string) =>
+                PerpEventLayout.decode(Buffer.from(event, 'base64'))
+
+            if (isSnapshot) {
+                for (const event of data.events.map(parseEvent)) {
+                    const fill = perpMarket.parseFillEvent(event.fill)
+
```
Comment (mschneider) [inline src/scripts/xmm.ts:239]: a bit annoying we don't have the slot number here

File: src/scripts/xmm.ts (hunk 3 above)
Comment (waterquarks) [reply src/scripts/xmm.ts:261]: Agreed, this is a much simpler solution

File: src/scripts/xmm.ts (hunk 9 above)
Comment (waterquarks) [reply src/scripts/xmm.ts:239]: Yeah having slot number for each fill in the snapshot message would be handy in this case - removed this bit in the meantime

File: src/scripts/xmm.ts (hunk 6 above)
Comment (waterquarks) [reply src/scripts/xmm.ts:109]: Looked at the code for `getMultipleAccounts` - no need for this check as the context is the same for all elements in the result array; removed it

File: src/scripts/xmm.ts (hunk 5 above)
Comment (waterquarks) [reply src/scripts/xmm.ts:407]: Re-enabled, was just fixing compiler errors

**Review by mschneider — APPROVED** [review]

**Pull request merged**